forem/forem
# Add the Ability to Remove Badge Achievements From Users

## What type of PR is this? (check all applicable)


## Description
This PR gives Admins the ability to remove badges from users and separates `badge_achievement`-related code from `badge`-related code. To accomplish this, this PR does the following:
- Adds an `Admin::BadgeAchievement::Controller`
- Adds `#index` and `#destroy` actions to the `Admin::BadgeAchievements::Controller` and ports over `#award` and `#award_badges` from the `Admin::Badges::Controller`
- Removes `#award` and `#award_badges` from the `Admin::Badges::Controller`
- Adds routes for `badge_achievements` and redirects to ensure that Admins to not encounter errors potentially caused by bookmarking the old `badge_achievement` routes
- Adds an `index` and `award` view for Badge Achievements
- Adds a warning to the `index` view stating: "Note: Badges that are automatically rewarded cannot and will not be removed from users!"
- Adds a search by `user_id` bar to the `index` view
- Limits the number of badges displayed on the `index` view to 15, order by descending `created_at`
- Adds specs around awarding and deleting `BadgeAchievements`
- Removes `badge_achievement`-related tests from the `badges_spec.rb`
- Adds a Badge Achievements menu item to the Admin nav bar

## Related Tickets & Documents
Relates to a private issue.

## QA Instructions, Screenshots, Recordings
**To QA these changes, first ensure that you are a `super_admin` with badges seeded, then please do the following:**
- Navigate to `/admin/badge_achievements` by typing out the URL or by navigating to `/admin` and then clicking the "Badge Achievements" navigation tab:
![Screen Shot 2020-08-21 at 12 22 43 PM](https://user-images.githubusercontent.com/32834804/90922339-1173c780-e3a9-11ea-94e8-f7392c505ac0.png)
- Click the "Award Badges" button to navigate to the newly moved badge award form and click the "Back to Badge Achievements" button to return to the `index` page:
![Screen Shot 2020-08-20 at 11 43 20 AM](https://user-images.githubusercontent.com/32834804/90806632-83311000-e2da-11ea-99b6-b99a2d50c4b3.png)
- Once you're back on the `index` view, test out the search by `user_id` functionality by searching for a user who has a Badge Achievement by ID:
![Screen Shot 2020-08-21 at 12 23 48 PM](https://user-images.githubusercontent.com/32834804/90922415-35cfa400-e3a9-11ea-9cff-662c8ec05f4c.png)
![Screen Shot 2020-08-21 at 12 23 59 PM](https://user-images.githubusercontent.com/32834804/90922421-37996780-e3a9-11ea-84ec-870ff2686729.png)
- Finally, remove a badge from a user (yourself, another user, or both!). You will be redirected back to the `index` view and
 should see a pop-up confirming the removal and then a confirmation message of the removal after clicking "Remove":
![Screen Shot 2020-08-21 at 12 24 29 PM](https://user-images.githubusercontent.com/32834804/90922475-4a13a100-e3a9-11ea-96d1-2c578219b6a4.png)
![Screen Shot 2020-08-21 at 12 24 38 PM](https://user-images.githubusercontent.com/32834804/90922482-4d0e9180-e3a9-11ea-9304-4d18df87d271.png)
- To ensure that everything worked properly, navigate to the user's page and double-check that the badge has been removed from their profile:
![Screen Shot 2020-08-21 at 12 25 32 PM](https://user-images.githubusercontent.com/32834804/90922541-6b748d00-e3a9-11ea-8dd9-9f3fa0ff7597.png)
- Try to break things! 🛠️ 

## Added tests?

- [x] yes, and refactored existing specs

## Added to documentation?


## [optional] Are there any post-deployment tasks we need to perform?
N/A

## [optional] What gif best describes this PR or how it makes you feel?

![backstreet boys](https://media.giphy.com/media/hFXwY4lER3oBO/giphy.gif)
